### ISSUE TITLE
Manipulation with window location is no longer possible

### ISSUE THREAD

**Issue by avlapeto** [jsdom/jsdom#3492]
### Basic info:

- **Node.js version:** 18.12.0
- **jsdom version:** 21.0.0

### Minimal reproduction case

We have a page that performs a redirect to an external page based on query string parameters:
```html
...
<script>
const redirectUrl = 'https://github.com/'
window.location.href = redirectUrl;
// window.location.assign(redirectUrl);
</script>
...
```
and integration test that expects that `location.href` will be changed (or `location.assing` will be invoked with some params).
```js
    const windowLocation = {};
    const { window } = new JSDOM(response.text, {
      beforeParse(window) {
        delete window.location;
        window.location = windowLocation;
      },
      runScripts: 'dangerously',
    });
   expect(windowLocation.href).toBe('https://github.com/');

```
Previously (jsdom 20.0.3) we changed the original location object to a syntetic one and ran an expectation on it. But since `window.location` became non-configurable, it's no longer possible. 
Any advice about options to spy on location object?

Right now, we are getting:
```
Error: Not implemented: navigation (except hash changes).
```

Thanks in advance!

**Comment by jenseng** [comment]
So this won't work for all cases, but if you can control which "window" is used by your code under test, you can inject a more test-friendly version (as in [this react-router PR](https://github.com/remix-run/react-router/pull/10453/files#diff-c9935ad9529fdeccc3a3e43718e797661372ef59a312debc3fb7cec2801314d1R432-R440) to upgrade jsdom).

e.g. given some app code we want to test:

```javascript
function goSomewhere({ redirectUrl }) {
  window.location.assign(redirectUrl);
}
```

If we change it like so:

```javascript
function goSomewhere({ redirectUrl, window: w = window }) {
  w.location.assign(redirectUrl);
}
```

Existing callers callers will still default to the real window. But in our test code we can then do something like this:

```javascript
const testWindow = { ...window, location: { ...window.location, assign: jest.fn() } };
goSomewhere({ redirectUrl: "https://github.com", window: testWindow });
expect(testWindow.location.assign).toHaveBeenCalledWith("https://github.com");
```

**Comment by avlapeto** [comment]
@jenseng 
Thanks much for your suggestion! Based on it I improved my tests to work with the newest version of `jest`. Unfortunately it's not possible in our case to extract function declaration to a js file and easily test it.

My solution is:

```html
...
<script>
// ... Business logic start
// ... Business logic end

const redirectUrl = 'https://github.com/'
window.redirect = window.redirect || function (redirectUrl, win) {
    win.location.href = redirectUrl;
  }
redirect(redirectUrl, window);
</script>
...
```

And test it in a way:
```js
const { window } = new JSDOM(response.text, {
  beforeParse(window) {
    window.redirect = jest.fn();
  },
  runScripts: 'dangerously',
});
expect(window.redirect).toHaveBeenCalledWith(redirectParam, window);;
```

It's note actually check that redirect is happening, but it test business logic before redirect should happen.

From my point of view this issue could be closed, since with propper decomposition you could test window object.

**Comment by uladzimirdev** [comment]
great solution @jenseng, unfortunately if you do not use window directly, it can be tricky to override window. Solution which worked in my case I found [here](https://github.com/jestjs/jest/issues/5124#issuecomment-914606939).

```ts
/**
 * Changes the URL used by the global JSDOM instance in the current window.
 *
 * NOTE: This does not change the origin which would result in a security exception.
 */
function changeJSDOMURL(url) {
  const newURL = new URL(url);
  const href = `${window.origin}${newURL.pathname}${newURL.search}${newURL.hash}`;
  history.replaceState(history.state, null, href);
}
```

**Comment by donaldpipowitch** [comment]
My strategy now: we moved all `location.href` assignments to a `setLocationHref` call that is imported from a module, so we can easily mock it in Jest. Not nice, but works.

**Comment by romulof** [comment]
Did anyone ever consider creating run time api for mocking?

Something like:
```ts
window.location._jsDomMock(() => ({
  ...window.location._jsDomOriginal,
  search: '?/potato=true',
}));
```

**Comment by theKashey** [comment]
removing [location: { configurable: false },](https://github.com/jsdom/jsdom/commit/0c8eb3a35973c7e95af9b1bdc81e5af2cc629924#diff-b5cd5c96785357dc930f47c18b45d1626b467e8c16068720a31c0cfc0d8344d3R482) is the only solution working for me.

**Comment by Daniel15** [comment]
> removing [location: { configurable: false },](https://github.com/jsdom/jsdom/commit/0c8eb3a35973c7e95af9b1bdc81e5af2cc629924#diff-b5cd5c96785357dc930f47c18b45d1626b467e8c16068720a31c0cfc0d8344d3R482) is the only solution working for me.

This is what I ended up doing too. We had a large number of tests (over 1000) that mock `window.location`, each in subtly different ways, so it would have been non-trivial to update them all to mock it a different way. I ended up using [patch-package](https://yarn.pm/patch-package) to make everything configurable again:
```patch
diff --git a/node_modules/jsdom/lib/jsdom/browser/Window.js b/node_modules/jsdom/lib/jsdom/browser/Window.js
index 247acad..45a3db9 100644
--- a/node_modules/jsdom/lib/jsdom/browser/Window.js
+++ b/node_modules/jsdom/lib/jsdom/browser/Window.js
@@ -477,10 +477,10 @@ function Window(options) {
     event: makeReplaceablePropertyDescriptor("event", window),
 
     // [LegacyUnforgeable]:
-    window: { configurable: false },
-    document: { configurable: false },
-    location: { configurable: false },
-    top: { configurable: false }
+    window: { configurable: true },
+    document: { configurable: true },
+    location: { configurable: true },
+    top: { configurable: true }
   });
 
 
diff --git a/node_modules/jsdom/lib/jsdom/living/generated/Location.js b/node_modules/jsdom/lib/jsdom/living/generated/Location.js
index fc4d1dd..c855bd5 100644
--- a/node_modules/jsdom/lib/jsdom/living/generated/Location.js
+++ b/node_modules/jsdom/lib/jsdom/living/generated/Location.js
@@ -322,19 +322,19 @@ function getUnforgeables(globalObject) {
       }
     });
     Object.defineProperties(unforgeables, {
-      assign: { configurable: false, writable: false },
-      replace: { configurable: false, writable: false },
-      reload: { configurable: false, writable: false },
-      href: { configurable: false },
-      toString: { configurable: false, writable: false },
-      origin: { configurable: false },
-      protocol: { configurable: false },
-      host: { configurable: false },
-      hostname: { configurable: false },
-      port: { configurable: false },
-      pathname: { configurable: false },
-      search: { configurable: false },
-      hash: { configurable: false }
+      assign: { configurable: true, writable: false },
+      replace: { configurable: true, writable: false },
+      reload: { configurable: true, writable: false },
+      href: { configurable: true },
+      toString: { configurable: true, writable: false },
+      origin: { configurable: true },
+      protocol: { configurable: true },
+      host: { configurable: true },
+      hostname: { configurable: true },
+      port: { configurable: true },
+      pathname: { configurable: true },
+      search: { configurable: true },
+      hash: { configurable: true }
     });
     unforgeablesMap.set(globalObject, unforgeables);
   }

```

**Comment by sergeyshmakov** [comment]
Had the same issue while trying to make `jest.spyOn(window, "location", "get")` work.
Based on @theKashey suggestion I created `JsdomHelper` to deal with it as accurate as possible:

```typescript
export class JsdomHelper {
    static unlockProperty = <T extends object>(
        object: T,
        key: keyof T,
        descriptor: PropertyDescriptor & ThisType<T> = {}
    ) => {
        Object.defineProperty(object, key, {
            configurable: true,
            ...(descriptor.get ? descriptor : {
                value: {...object[key]},
                ...descriptor
            })
        });
    };

    static unlockWindow = () => {
        this.unlockProperty(globalThis, "window");
    };
}
```

And then used it in my `jest` mocks like this:

```typescript
export const spyLocation = () => {
    JsdomHelper.unlockWindow();
    JsdomHelper.unlockProperty(window, "location", {
        get() {
            return this.location;
        }
    });
    return jest.spyOn(window, "location", "get") as SpyInstance<Partial<Location>>;
};
```

The final solution to use it in tests now looks like:

```typescript
describe("SomeApi tests", () => {
    const someApi = new SomeApi();
    const locationSpy = spyLocation();

   describe("Test api on specific url", () => {
        beforeAll(() => {
            locationSpy.mockReturnValue({pathname: "/some-specific-url"});
        });
        // Test cases here
    });
});
    
```

**Comment by createthis** [comment]
This is bullshit. I shouldn't have to monkey patch jsdom just to mock window.location. The security nerds have gone too far.

**Comment by isocroft** [comment]
Well @createthis , i do appreciate your candor and enthusiasm.

Yes, you are right: 

>one shouldn't have to monkey patch **JSDOM** just to mock `window.location`.

But, **JSDOM** is a half-decent, impressive and convincing imitation of a web browser and does not and cannot support URL navigation within a Node.js process in a way that makes much sense (or is desirable) in the context of just **JSDOM**

Yet, in the context of being part of a test environment artefact within something like **Jest** it becomes desirable. But **JSDOM** does not provide a plugin setup that allows client to modify the way parts of **JSDOM** works.

Therefore, a client has no option than to monkey patch it.

**Comment by sibelius** [comment]
do you accept a pull request modifying location to be like this

```
location: { configurable: true },
```

as jsdom is mostly used by test environment, it makes sense to be able to modify and configure location to simplify tests

**Comment by domenic** [comment]
No, we don't accept pull requests that make us non-standards-compliant.

**Comment by cpojer** [comment]
@domenic we are going to ship Jest 30 soon with jsdom 26. People are angry (with me?) about this, and jsdom's dominant use case is as part of Jest. Would you be open to making the behavior of the `location` prop (and/or other props on window) configurable?

The default would stay the same, and we'd just pass the flag to make the fields configurable. That way you don't compromise on standards compliance, while also supporting one of the major use cases of jsdom.

If you agree, I'll find somebody to make the PR, and all we need is a new jsdom 26 patch release. What do you say?

**Comment by domenic** [comment]
No, sorry, I'm not interested in supporting that use case in jsdom.

**Comment by answer-equals-42** [comment]
@domenic So what is the alternative if I want to make sure that the window.location.href has been changed or window.open has been called etc.? I mean everyone has to use some hacks to make this work. Would it not make sense to have something that makes it more user-friendly? Honestly, this is a bit of a let down...

**Comment by domenic** [comment]
I don't know the answer to that. You should research how you would do that in a real browser, and then apply the code that you find in jsdom.

**Comment by answer-equals-42** [comment]
@domenic I don't understand the decision and the change. Would you mind elaborating why this change is necessary? I think @cpojer makes a perfectly valid point here. Why can the consumers of the library not decide on their own if they want to opt-in or opt-out of these changes? Usually, if there is a change like this there is a "better", more modern, more secure alternative, but removing the functionality entirely kind of misses the point as users still want to test this use-case.

... well in a real browser I end up using window.open for example. So what else can I do? I'll have to wrap all my window.open calls in a different function in order to mock it properly and track it which results in breaking changes.

**Comment by domenic** [comment]
> don't understand the decision and the change. Would you mind elaborating why this change is necessary?

I've explained it several times. To match browsers, which is the goal of the jsdom project.

> Why can the consumers of the library not decide on their own if they want to opt-in or opt-out of these changes?

Consumers who want a behavior besides matching browsers are free to user a different library besides jsdom. It's not a configuration we're interested in supporting.

**Comment by Daniel15** [comment]
I understand both perspectives. jsdom wants purity. It aims to be standards compliant and only allow things that are allowed by the DOM specs. It wants to behave identically to how a real production browser behaves. 

On the other hand, Jest does a lot of things you normally *wouldn't* do in production code. Things like monkey patching, swapping out real modules for shims, etc. Jest intentionally isn't a real production environment - it's more like a sandbox where you have a lot more control over everything. 

The two designs won't always see eye to eye... jsdom wants to be as strict as possible, while Jest is essentially the opposite and wants to be as flexible as possible. 

If jsdom won't budge on this (even to add it as an option that's disabled by default), then maybe Jest should allow the DOM implementation to be swapped out for a different one? jsdom could be forked (which isn't great). 

Perhaps Jest can monkey patch jsdom to do what's needed. That seems fragile though. The Jest documentation could explain how to use patch-package to patch jsdom (and even provide the patch), but that's not great either.

**Comment by answer-equals-42** [comment]
@domenic Well I get your point with being standard compliant but IMO the argument is quite weak as it does not solve the problem at hand - after all the jsdom is supposed to be a mocking tool not a "real world" tool if I want to run it on a real browser I'd use e2e testing tools like cypress or wdio. It feels like by sticking to this ideology you kind of fight the purpose why this library exists in the first place. The problem is that there is no good alternative (at least that I know of) for unit and component integration test with regards to window which is an important part of JS. Yes, I can ditch jsdom - but why should I do that? Adding another dependency/lib just for that purpose seems to be overkill. Why not just support both options?

**Comment by domenic** [comment]
> jsdom is supposed to be a mocking tool

This is not correct.

**Comment by answer-equals-42** [comment]
> > jsdom is supposed to be a mocking tool
> 
> This is not correct.

@domenic Sorry. I was a bit imprecise there. It simulates a browser-like environment in Node. It’s not a mocking tool in the traditional testing sense, but it mimics browser APIs, which can be seen as a kind of environment-level mock or simulation. Often used together with jest for example which then limits what you can actually do.

**Comment by isocroft** [comment]
Just so i am clear, what is so wrong about monkey-patching the location object for all use cases in user-land (i.e. Jest e.t.c) ? 

Especially since @domenic isn't budging on this.

Please, someone help me understand.

**Comment by Daniel15** [comment]
That's the issue - Monkey-patching the location object is not possible because it's marked as non-configurable.

**Comment by isocroft** [comment]
Yes @Daniel15 , directly monkey-patching it is not possible yet i think patching it indirectly (by wrapping it in a shell of indirection) is possible.

I am also indirectly patching `window.print(...)` and `window.matchMedia(...)` here in [my **mocklets** library](https://github.com/codesplinta/mocklets/blob/main/.export/jest.setup.js#L1-L244) while silencing the _"Not implemented"_ error from JSDOM (Watch this short video to understand why i had to patch `window.print(...)`: https://www.youtube.com/watch?v=JRfJko66QOo&t=150s).

See further/same here for `window.location`:  [**mocklets** library](https://github.com/codesplinta/mocklets/blob/main/index.js#L70-L193). I am also about to include history updates too.

**Comment by vitorelourenco** [comment]
I think the maintainers get to have their takes on how they conduct the repo. Also allowing non compliance as a configurable thing might lead to issues and PRs that attempt to modify upstream behavior based on an even further downstream non-compliant configuration need. I'm not well versed in JSDOM but I work with people who are facing issues with jest because of it (and that makes it an issue for everybody as a result). Is there an easy path for a jsdom"-extra" build? So that what is actually consumed by jest is not jsdom"-pure" but an alternative build that does allow for such things?

**Comment by marcellobarile** [comment]
how would you assert that `window.location.assign` was called with an expected value after this change?

**Comment by isocroft** [comment]
@marcellobarile . Glad you asked.

`window.location.assign` and all other functions like `window.location.replace` and `window.location.reload` are wrapped with `jest.fn(...)` when being patched in [**mocklets** library](https://github.com/codesplinta/mocklets/blob/main/index.js#L70-L193)

**Comment by maury91** [comment]
> You should research how you would do that in a real browser, and then apply the code that you find in jsdom.

In the real world if I want to change the location I would use one of the many methods like `assign` or `replace`, or even a simple `location.href = x`.

The problem is that JSDom just spits out a "not implemented" warning.

I understand wanting to be as compliant as possible, but if the location is locked and the methods that can change it are not implemented, testing any code with navigation in JSDom becomes impossible without hacks.

Cause one road is blocked because it will make JSDom less standard-compliant, we could look into the other road, implementing what is missing.

@domenic  Would you be willing to accept a PR that implements a basic navigation system?

**Comment by marcellobarile** [comment]
> how would you assert that window.location.assign was called with an expected value after this change?

follow-up:
For the time being, I had to create a helper that hides the interaction with the browser's low-level API (location.assign, etc) and spy/mock it in my tests instead.

It's a bummer that I had to touch production code to cope with a dev dependency though.

**Comment by isocroft** [comment]
>Cause one road is blocked because it will make JSDom less standard-compliant, we could look into the other road, implementing what is missing.

This is a welcome idea. All we have now are hacks. I support it

**Comment by domenic** [comment]
> [@domenic](https://github.com/domenic) Would you be willing to accept a PR that implements a basic navigation system?

If you can implement a spec-compliant version of https://html.spec.whatwg.org/#navigation-and-session-history, please do. However, note that it's extremely complicated. I'd suggest first putting together a design doc, instead of a pull request. As a small taste of the issues you'd be raising, you'd need to first separate out `Window` and `WindowProxy`, so that navigation changes the `Window` but leaves the `WindowProxy` the same. I'm not sure that's possible in Node.js today, but if you have a solution written up, I'd be interested to review it.

**Comment by Netail** [comment]
What about exporting 2 JSDOM versions from the package, one fully browser compliant (a strict version) another fully mockable (a loose version)

`import { JSDOM, MockableJSDOM } from "jsdom";`

**Comment by domenic** [comment]
I support someone else creating a fork of jsdom with their own modifications. I don't have plans to maintain two versions in my own limited time.

**Comment by PenghaiZhang** [comment]
I guess the JSDOM team should do a survey to know what roles use this library mostly for what purposes. If the majority users use this library for an intention that is NOT what this library is designed for, it would be important to reconsider those users' requests. Or maybe this product is the lowest priority in your organisation and you will not spend time on it, then I can only hope there is proper alternative coming shortly.

**Comment by PenghaiZhang** [comment]
> > jsdom is supposed to be a mocking tool
> 
> This is not correct.

But this is could be how most users use it in reality.

**Comment by cpojer** [comment]
While I don't like this decision by the `jsdom` team either, @PenghaiZhang I think you have a fundamental misunderstanding of open source, and the history here. From my perspective, I agree that these days jsdom is mostly used in a context for emulating a browser environment for testing, but the authors of `jsdom` did not originally build it for that purpose and since it is open source, they were not in control of us (the Jest team) deciding to use it. Where they want to take the open source project in the future is entirely up to them, and if you don't like it, you can fork it and build the tool you need yourself.

**Comment by PenghaiZhang** [comment]
> While I don't like this decision by the `jsdom` team either, [@PenghaiZhang](https://github.com/PenghaiZhang) I think you have a fundamental misunderstanding of open source, and the history here. From my perspective, I agree that these days jsdom is mostly used in a context for emulating a browser environment for testing, but the authors of `jsdom` did not originally build it for that purpose and since it is open source, they were not in control of us (the Jest team) deciding to use it. Where they want to take the open source project in the future is entirely up to them, and if you don't like it, you can fork it and build the tool you need yourself.

Sorry mate but I don't see any misunderstanding. Firstly, like what you said, how this library is used in reality nowadays possibly does not its original intention. Hence, it is reasonable for the authors to say no. But, given the fact that there have been so many people requesting/discussing this kind of feature (not just here, but over in Jest and more) , on one hand the authors should be proud that their product is widely used, on the other hand if they do care their end users' feedback, they should take a certain action rather than simply refusing everything. 

Lastly, I absolutely understand what open source is, hence I said I could only hope someone else would make an alternative. Why do I not do this by myself ? The answer is simple. Because I do not use this library directly. It is used in Jest and if Jest had gone with a different one I would not have been here.

**UPDATE**:

I am now trying to use `@jest/environment-jsdom-abstract`.


**Update:**

The two purposes are clearly stated:

![Image](https://github.com/user-attachments/assets/533c98b4-a235-4d3d-8fdb-449b952598ba)

And the ease of testing comes first. 🤷 🤕 🤦

**Comment by guilhermesimoes** [comment]
Hey, I work at Sky and this is a very common use-case for us. We have hundreds of tests that do `jest.spyOn(window, 'window', 'get')` or `Object.defineProperty(window, 'location', ...)`.

We do this to:
1. Simulate certain window APIs that only exist on TVs (for example, Samsung TVs have `window.tizen.package.getPackageInfo()`, LG TVs have `window.webOS.deviceInfo(...)`, etc)
2. Test certain interactions with the TV app URL

We'd really like to upgrade to jsdom v26.1.0, to stop including the bloated `form-data` dependency (nice work on #3850, @domenic!), but this breaking change is making this difficult. I'd really like to avoid patching our test framework or changing production code.

@domenic please don't take this personally, I don't intend this as an attack 🙏 but just like that individual is very inflexible in his quest to support node 0.4, to the detriment of devs worldwide, I do think you're being a bit inflexible in your quest for "jsdom purity", to the detriment of current users of this package.

Just look at the amount of issues and PRs that reference this issue, above this comment of mine. This seems like a very common use-case.
